src/porepy/numerics/nonlinear/nonlinear_solvers.py

```diff
@@ -87,6 +88,7 @@ def newton_step() -> None:
             nonlocal nonlinear_increment
             nonlocal reference_residual
             nonlocal residual_norm
+            nonlocal residual
```

Comment:
Why is this needed now and not previously? Same question for line 73.

Reply:
Now it may be the case, that the residual is not recomputed. In this case, the value outside of the internally defined `newton_step` needs to be forwarded, this would be then the default value. An alternative (maybe cleaner?) could be to remove line 73 and 91, and instead set residual to `None` when the computation is not needed, i.e., in an `else` case after line 112. I actually would opt for that solution. Any opinion?

Reply:
I see. Yes, I think I agree that the else option is better.

Reply:
Just to check my understanding: Would it make sense to move the line 112 call to inside check_convergence?